client/network/src/protocol.rs

```diff
@@ -371,6 +374,35 @@ where
 			sc_peerset::Peerset::from_config(sc_peerset::PeersetConfig { sets })
 		};
 
+		if let Some(loaded_peerinfo) = network_config
+			.net_config_path
+			.as_ref()
+			.map(persist_peers::peersets_load)
+			.transpose()?
+		{
+			use sc_peerset::{ReputationChange, SetId};
+
+			for (peer_id, reputation, sets) in loaded_peerinfo {
+				log::info!(
+					"Restoring {:?} with sets: {:?}, reputation: {:?}",
+					peer_id,
+					sets,
+					reputation
+				);
+				peerset.report_peer(
```

Comment:
Was this also used in the old implementation?

Can we not just use the swarm here: https://docs.rs/libp2p/latest/libp2p/struct.Swarm.html#method.add_external_address

Reply:
We probably should hook in [here](https://github.com/paritytech/substrate/blob/00cc5f104176fac6f5a624bced22a2192c7c0470/client/network/src/discovery.rs#L249-L281) or not?

Reply:
I am not sure I understand the question about the "old implementation".


> We probably should hook in [here](https://github.com/paritytech/substrate/blob/00cc5f104176fac6f5a624bced22a2192c7c0470/client/network/src/discovery.rs#L249-L281) or not?

I wanted to populate the `peerset` with the previously known peers before anything else happens with it. To guarantee that I populated it right after it is created.

Reply:
> I am not sure I understand the question about the "old implementation".

As the issue mentioned, this feature was already implemented once. By "old implementation" I mean the one that was implemented back then. https://github.com/paritytech/substrate/blob/d270ae4f507b14a92ceca3bd54dc2d9c260607f6/core/network-libp2p/src/custom_proto/topology.rs

> I wanted to populate the `peerset` with the previously known peers before anything else happens with it. To guarantee that I populated it right after it is created.

If you read the old implementation you see that it stores all nodes and we used these nodes on restart to bootstrap the network. Storing only the peers that are in the peerset is probably not enough. There are also peersets in Polkadot where it wouldn't make sense to boostrap them from their old state because they depend on the current set of active validators for example.